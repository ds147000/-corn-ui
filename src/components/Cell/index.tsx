import React, { useCallback } from 'react'
import ClassNames from 'classnames'
import { View, Text } from '@tarojs/components'
import { ViewProps } from '@tarojs/components/types/View'
import Icon from '../Icon'
import Link, { LinkProps } from '../Link'
import CellList from './list'

export interface CellProps extends ViewProps, LinkProps {
  /** 标题 */
  label?: string | React.ReactNode
  /** 内容 */
  value?: string | React.ReactNode
  /** 内容为空占位提示 */
  placeholder?: string | React.ReactNode
  /** 是否显示箭头 */
  arrow?: boolean
  /** 尾缀内容 */
  suffix?: React.ReactNode | string
  /** 禁用 */
  disabled?: boolean
  /** 跳转链接，填写启用A链接行为。同时会显示箭头 */
  to?: string
}

type CellView =  { List: typeof CellList }



const Cell: React.FC<CellProps> & CellView = ({
  label, value, arrow, suffix,
  disabled, to, children, onClick,
  className, placeholder, ...props
}) => {

  const _onClick = useCallback((event): void => {
    if (disabled) return
    onClick?.(event)
  }, [ onClick, disabled ])

  const _props: ViewProps = {
    className: ClassNames('xrk-cell xrk-f xrk-ac', disabled && 'xrk-cell-disable', className),
    onClick: _onClick,
    ...props
  }

  const showArrow = (arrow || Boolean(to)) && arrow !== false
  const showPlaceholder = Boolean(value) === false && Boolean(children) === false

  const content = (
    <>
      <View className="xrk-cell-label">{label}</View>
      <View className="xrk-cell-value">
        {value}{children}
        {showPlaceholder && <Text className="xrk-cell-placeholder">{placeholder}</Text>}
      </View>
      <View className="xrk-cell-suffix xrk-f xrk-ac">
        {suffix}
        {showArrow && <Icon name="big-right" data-testid="arrow" />}
      </View>
    </>
  )

  if (to)
    return (<Link {..._props} to={to} disabled={disabled} target="View" >{content}</Link>)

  return (<View {..._props} >{content}</View>)
}

Cell.List = CellList

export default Cell
