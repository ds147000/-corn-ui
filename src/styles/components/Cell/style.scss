.corn-cell {
  width: 100%;
  min-height: $cellHeight;
  background-color: $white;
  padding: 20px $pad;
  box-sizing: border-box;
  font-size: $cellSize;
  line-height: $cellSize;
  transition: opacity 0.1s;

  &:active {
    opacity: 0.7;
  }

  &.corn-cell-disable {
    background-color: $disable;
    opacity: 0.3;

    &:active {
      background-color: $disable;
      opacity: 0.3;
    }
  }

  &-label {
    width: $cellWidth;
    font-weight: 400;
    color: $text2;
  }

  &-value {
    flex: 1;
    color: $text1;
    font-weight: bolder;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  &-suffix {
    color: $text2;

    .corn-icon {
      color: $cellIconColor;
    }
  }
}

.corn-cell-list-line {
  >.corn-cell {
    border-bottom: 2px solid $bg;

    &:last-child {
      border-bottom: none;
    }
  }
}
