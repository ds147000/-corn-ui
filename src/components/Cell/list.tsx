import React, { useMemo } from 'react'
import ClassNames from 'classnames'
import { View } from '@tarojs/components'
import { ViewProps } from '../../types/View'

export interface CellListProps extends ViewProps {
  /** 是否显示边框 */
  line?: boolean
}

const CellList: React.FC<CellListProps> = ({ children, line, className, ...props }) => {
  const _class = useMemo(() => {
    return ClassNames(line && 'corn-cell-list-line', className)
  }, [ line, className ])

  return (
    <View {...props} className={_class}>
      {children}
    </View>
  )
}

export default CellList
