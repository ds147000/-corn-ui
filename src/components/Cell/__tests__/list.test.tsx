import React from 'react'
import { render } from '@testing-library/react'
import CellList from '../list'
import Cell from '../index'

describe('Cell.List', () => {

  test('basis', () => {
    const screen = render(
      <CellList>
        <Cell label="限时奖励" />
        <Cell label="送葵花籽" />
      </CellList>
    )

    expect(screen.container).toMatchSnapshot()
  })

  test('line', () => {
    const screen = render(
      <CellList line >
        <Cell label="限时奖励" />
        <Cell label="送葵花籽" />
      </CellList>
    )

    expect(screen.container).toMatchSnapshot()
  })
})


