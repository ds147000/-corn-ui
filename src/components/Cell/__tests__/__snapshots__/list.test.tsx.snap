// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell.List basis 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="corn-cell corn-f corn-ac"
    >
      <div
        class="corn-cell-label"
      >
        限时奖励
      </div>
      <div
        class="corn-cell-value"
      >
        <span
          class="taro-text corn-cell-placeholder"
        />
      </div>
      <div
        class="corn-cell-suffix corn-f corn-ac"
      />
    </div>
    <div
      class="corn-cell corn-f corn-ac"
    >
      <div
        class="corn-cell-label"
      >
        送葵花籽
      </div>
      <div
        class="corn-cell-value"
      >
        <span
          class="taro-text corn-cell-placeholder"
        />
      </div>
      <div
        class="corn-cell-suffix corn-f corn-ac"
      />
    </div>
  </div>
</div>
`;

exports[`Cell.List line 1`] = `
<div>
  <div
    class="corn-cell-list-line"
  >
    <div
      class="corn-cell corn-f corn-ac"
    >
      <div
        class="corn-cell-label"
      >
        限时奖励
      </div>
      <div
        class="corn-cell-value"
      >
        <span
          class="taro-text corn-cell-placeholder"
        />
      </div>
      <div
        class="corn-cell-suffix corn-f corn-ac"
      />
    </div>
    <div
      class="corn-cell corn-f corn-ac"
    >
      <div
        class="corn-cell-label"
      >
        送葵花籽
      </div>
      <div
        class="corn-cell-value"
      >
        <span
          class="taro-text corn-cell-placeholder"
        />
      </div>
      <div
        class="corn-cell-suffix corn-f corn-ac"
      />
    </div>
  </div>
</div>
`;
